source/httpfs/driver.go

```diff
@@ -37,16 +24,35 @@ type Driver struct {
 // this driver. This reduces the number of error handling branches in clients of
 // this package without compromising on correctness.
 func New(fs http.FileSystem, path string) source.Driver {
-	var d Driver
+	var d driver
 	if err := d.Init(fs, path); err != nil {
 		return &failedDriver{err}
 	}
 	return &d
 }
 
-// Init prepares not initialized Driver instance to read migrations from a
+// Open is part of source.Driver interface implementation. It always returns
+// error because http.FileSystem must be provided by the user of this package
+// and created using WithInstance() function.
+func (d *driver) Open(url string) (source.Driver, error) {
+	return nil, errors.New("not implemented")
+}
+
+// Migrator is a helper service for creating new source drivers working with
+// http.FileSystem instances. It implements all source.Driver interface methods
+// except for Open(). New driver could embed this struct and add missing Open()
+// method.
+//
+// To prepare Migrator for use Init() function.
+type Migrator struct {
```

Comment:
I am trying to find an alternative name to avoid package and struct stuttering (where package name is the same as structure name).

I totally agree with you that `Migrator` is not a good one. What do you think about `Source`? This struct will only be used for embedding into other drivers. For other packages it could only be referenced by its full name `httpfs.Source`. The usage would look like:

```
type mydriver {
    // private fields
    httpfs.Source
}
```

Reply:
Considerations for names for the embedded struct:
* ~`httpfs.Migrator`~ - We've agreed that this isn't a good name
* `httpfs.HTTPFS` - I'm ok with the stutter, but agree that we should be able to come up with a more descriptive name
* `httpfs.Source` - Not an accurate name since it doesn't fully implement a source driver and is not a source on its own
* `httpfs.Embedded` - Not fully descriptive. e.g. only talks about one aspect of the partial driver implementation
* `httpfs.EmbedMe` - Same issue as `httpfs.Embedded` and usage would be more awkward
* `httpfs.PartialDriver` - Seems descriptive enough. I don't think using `Partial` would be confused with a partial function due to the usage/context.
* `httpfs.PartialSource` - Also seems descriptive enough
* `httpfs.Partial` - Not as descriptive as `httpfs.PartialDriver` or `httpfs.PartialSource`

What do you think about `httpfs.PartialDriver` or `httpfs.PartialSource`?

Reply:
:+1: lets try `PartialDriver`.